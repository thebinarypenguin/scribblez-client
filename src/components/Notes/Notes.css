*,
*::before,
*::after {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  font-size: 16px;
  font-family:
    -apple-system,
    BlinkMacSystemFont,
    "Segoe UI",
    "Roboto",
    "Helvetica Neue", Arial, sans-serif,
    "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}

#site-container {
  width: 750px;
  margin: auto;
}


/* -------------------------------------------------------------------------- */

.note-list {
  list-style-type: none;
  padding: 1rem;
}

.note {
  border:1px solid lightgrey;
  padding: 0.25rem;
  margin-bottom: 1rem;
}

.note-meta {
  display: flex;
  justify-content: space-between;
  margin-bottom: 0.5rem;
}

.note-author {
  font-weight: bold;
}

/* -------------------------------------------------------------------------- */

#new-note-form {
  background-color: lightgrey;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  padding: 0.5rem;
}

#note-body {
  resize: none;
  width: 100%;
  flex-grow: 1;
}

#note-body,
#note-visibility,
#create-note {
  padding: 3px 6px;
  margin: 0 3px;
  height: 1.5rem;
}

.note-with-controls {
  display: flex;
}

.note-note {
  flex-grow: 1;
  border:1px solid lightgrey;
  padding: 0.25rem;
  margin-bottom: 1rem;
}

.note-controls {
  padding: 6px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  margin-bottom: 1rem;
}

.note-controls button {
  padding: 3px 6px;
  width: 100%;
}
