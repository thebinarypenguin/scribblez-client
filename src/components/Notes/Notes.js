import React from 'react';
import UserContext from '../../contexts/UserContext';
import TopBar from '../TopBar/TopBar';
import CreateNoteForm from '../CreateNoteForm/CreateNoteForm';
import NoteList from '../NoteList/NoteList';
import NotesService from '../../services/NotesService';

import './Notes.css';

class Notes extends React.Component {

  static contextType = UserContext;

  constructor(props) {
    super(props);

    this.state = {
      notes: [],
    };

    this.generateLinks = this.generateLinks.bind(this);
    this.createNote    = this.createNote.bind(this);
    this.editNote      = this.editNote.bind(this);
    this.deleteNote    = this.deleteNote.bind(this);
    this.saveNote      = this.saveNote.bind(this);
    this.cancelNote    = this.cancelNote.bind(this);
  }

  generateLinks() {

    let links = [];

    if (this.context.user) {

      links = [
        { name: 'Feed',     target: '/feed'     },
        { name: 'Notes',    target: '/notes'    },
        { name: 'Sign Out', target: '/sign-out' },
      ];

    } else {

      links = [
        { name: 'Feed',    target: '/feed'    },
        { name: 'Sign Up', target: '/sign-up' },
        { name: 'Sign In', target: '/sign-in' },
      ];
    }

    return links;
  }

  componentDidMount() {

    if (!this.context.user) {
      return this.props.history.push('/sign-in');
    }

    NotesService
      .getAllNotes()
      .then((notes) => {

        this.setState({ notes });
      })
  }

  createNote(payload) {

    NotesService
      .createNote(payload)
      .then((id) => {

        return NotesService
        .getAllNotes()
        .then((notes) => {

          this.setState({ notes });
        })
      });
  }

  saveNote(noteId, payload) {

    NotesService
      .updateNote(noteId, payload)
      .then((id) => {

        return NotesService
        .getAllNotes()
        .then((notes) => {

          this.setState({ notes });
        })
      });
  }

  deleteNote(noteId) {

    NotesService
      .deleteNote(noteId)
      .then((id) => {

        return NotesService
        .getAllNotes()
        .then((notes) => {

          this.setState({ notes });
        })
      });
  }

  editNote(noteId) {

    const newNotes = this.state.notes.map((n) => {
      if (n.id === noteId) {
        n._editable = true;
      } else {
        n._editable = false;
      }

      return n;
    });

    this.setState({notes: newNotes});
  }

  cancelNote() {

    const newNotes = this.state.notes.map((n) => {
      if (n._editable) {
        n._editable = false;
      }

      return n;
    });

    this.setState({notes: newNotes});
  }

  render() {

    return (
      <div id="site-container">

        <div id="notes-container">

          <TopBar links={this.generateLinks()} />

          <CreateNoteForm onSubmit={this.createNote}/>

          <NoteList
            notes={this.state.notes}
            onEdit={this.editNote}
            onDelete={this.deleteNote}
            onSave={this.saveNote}
            onCancel={this.cancelNote}
            />

        </div>

    </div>
    );
  }
}

export default Notes;
