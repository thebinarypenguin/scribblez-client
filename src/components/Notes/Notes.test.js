import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from "react-router-dom";
import renderer from 'react-test-renderer';

import Notes from './Notes';

it('renders without crashing', () => {

  const history = {
    push: () => {},
  };

  const container = document.createElement('div');

  const subject = (
    <MemoryRouter>
      <Notes history={history} />
    </MemoryRouter>
  );

  ReactDOM.render(subject, container);
  ReactDOM.unmountComponentAtNode(container);
});

it('renders the UI as expected', () => {

  const history = {
    push: () => {},
  };

  const subject = (
    <MemoryRouter>
      <Notes history={history} />
    </MemoryRouter>
  );

  const tree = renderer.create(subject).toJSON();
  expect(tree).toMatchSnapshot();
});
