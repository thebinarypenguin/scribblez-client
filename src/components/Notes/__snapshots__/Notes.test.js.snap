// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<div
  id="Notes"
>
  <div
    id="TopBar"
  >
    <div
      className="Container"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <h1>
          Scribblez
        </h1>
      </a>
      <ul>
        <li>
          <a
            href="/feed"
            onClick={[Function]}
          >
            Feed
          </a>
        </li>
        <li>
          <a
            href="/sign-up"
            onClick={[Function]}
          >
            Sign Up
          </a>
        </li>
        <li>
          <a
            href="/sign-in"
            onClick={[Function]}
          >
            Sign In
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="Container"
  >
    <form
      id="CreateNoteForm"
      onSubmit={[Function]}
    >
      <textarea
        id="note-body"
        placeholder="New note"
        rows="1"
      />
      <select
        id="note-visibility"
      >
        <option
          value="public"
        >
          public
        </option>
        <option
          value="private"
        >
          private
        </option>
      </select>
      <button
        id="create-note"
        type="submit"
      >
        Create
      </button>
    </form>
  </div>
  <div
    className="Container"
  >
    <ul
      id="NoteList"
    />
  </div>
</div>
`;
